model_compression_toolkit/gptq/keras/quantizer/soft_rounding/symmetric_soft_quantizer.py

```diff
@@ -0,0 +1,347 @@
+        """
+        Initializes a LinearTempDecay object.
+
+        Args:
+            t_max: maximal time step.
+            rel_start_decay: Decay step size at the beginning of the process.
+            start_b: Starting value of the regularization term.
+            end_b: Target value of the regularization term.
+        """
+
+        self.t_max = t_max
+        self.start_decay = rel_start_decay * t_max
+        self.start_b = start_b
+        self.end_b = end_b
+
+    def __call__(self, t: int) -> float:
+        """
+        Cosine annealing scheduler for soft quantizer regularization temperature term.
+
+        Args:
+            t: The current time step.
+
+        Returns: Scheduled temperature.
+        """
+
+        ind = tf.cast(t < self.start_decay, tf.float32)
+
+        rel_t = (t - self.start_decay) / (self.t_max - self.start_decay)
+
+        return self.start_b * ind + (1 - ind) * (
+                self.end_b + (self.start_b - self.end_b) * tf.math.maximum(0.0, (1 - rel_t)))
+
+
+class SymmetricSoftRounding(BaseTrainableQuantizer):
+    """
+    Trainable symmetric quantizer to optimize the rounding of the quantized values using soft quantization method.
+    """
+
+    def __init__(self, num_bits: int,
+                 per_axis: bool,
+                 signed: bool,
+                 power_of_two: bool,
+                 n_batches: int,
+                 quantization_parameter_learning: bool,
+                 threshold_values: np.ndarray,
+                 quantization_axis: int = -1,
+                 n_epochs: int = N_EPOCHS):
+        """
+        Initialize a SymmetricSoftRounding object with parameters to use
+        for the quantization.
+        Args:
+            num_bits: Number of bits to use for the quantization.
+            per_axis: Whether to quantize per-channel or per-tensor.
+            signed: Signedness to use for the quantization range.
+            power_of_two: Whether the threshold should be constrained or not.
+            n_batches: The expected number of batches for each trainig epoch.
+            quantization_parameter_learning: Whether to train the quantization threshold.
+            threshold_values: Threshold to use for the quantization.
+            quantization_axis: Axis of tensor to use for the quantization.
+            n_epochs: Number of epochs to run training for.
+        """
+
+        super().__init__()
+        self.num_bits = num_bits
+        self.per_axis = per_axis
+        self.signed = signed
+        self.power_of_two = power_of_two
+        self.quantization_parameter_learning = quantization_parameter_learning
+        self.quantization_axis = quantization_axis
+        self.threshold_shape = np.asarray(threshold_values).shape
+        self.threshold_values = np.reshape(np.asarray(threshold_values), [-1]) if self.per_axis else np.asarray(
+            threshold_values)
+        self.k_threshold = len(self.threshold_values) if self.per_axis else 1
```

Comment:
why "k_threshold" and not "num_channels"?

Reply:
Same reason